File: 2023/p9.ml
let read_file filename = 
  let ch = open_in filename in 
  let read_line () = 
    try Some (input_line ch) with End_of_file -> None 
  in 
  let rec read_lines lines = 
    match read_line () with 
    | None -> lines 
    | Some line -> read_lines (line :: lines)
  in 
  read_lines []

let line_to_seq line = 
  let seqs = String.split_on_char ' ' line in 
  List.map int_of_string seqs

let rec extrapolate seq =
  let rec diffs = function
    | [x; y] -> [y - x]
    | x :: (y :: ys as tl) -> (y - x) :: diffs tl
    | _ -> assert false
  in
  let last xs = List.nth xs (List.length xs - 1) in
  if List.for_all (fun n -> n = 0) seq then 0 else last seq + extrapolate (diffs seq)

let () =
  let lines = read_file "input_p9.txt" in 
  let seqs = List.map line_to_seq lines in 
  let ans1 = List.fold_left (fun acc s -> acc + extrapolate s) 0 seqs in 
  let ans2 = List.fold_left (fun acc s -> acc + extrapolate (List.rev s)) 0 seqs in 
  Printf.printf "Answer to part 1 is %d\nAnswer to part 2 is %d\n" ans1 ans2

File: 2023/p9.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> ReadFile(const string& filename) {
    ifstream is(filename);
    
    vector<string> lines;
    string s;
    while (getline(is, s)) {
        lines.push_back(s);
    }
    
    return lines;
}

vector<int> ParseSeq(const string& line) {
    vector<int> seq;
    
    istringstream is(line);
    int n;
    while (is >> n) {
        seq.push_back(n);
    }
    return seq;
}

int Extrapolate(const vector<int>& seq) {
   vector<vector<int>> seqs;
   seqs.push_back(seq);
   
   int i = 0;
   while (any_of(seqs[i].begin(), seqs[i].end(), [](int n) { return n != 0; })) {
       seqs.push_back(vector<int>(seqs[i].size() - 1));
       for (int j = 1; j < seqs[i].size(); j++) {
           seqs[i + 1][j - 1] = seqs[i][j] - seqs[i][j - 1];
       }
       i++;
   }
   
   for (; i > 0; i--) {
       seqs[i - 1].push_back(seqs[i - 1].back() + seqs[i].back());
   }
   return seqs[0].back();
}

int main() {
    vector<string> lines = ReadFile("input_p9.txt");
    vector<vector<int>> seqs;
    for (const string& line : lines) {
        seqs.push_back(ParseSeq(line));
    }
    
    int ans1 = 0;
    for (const vector<int>& seq : seqs) {
        ans1 += Extrapolate(seq);
    }
    printf("Answer to part 1 is %d\n", ans1);
    
    int ans2 = 0;
    for (const vector<int>& seq : seqs) {
        ans2 += Extrapolate(vector<int> (seq.rbegin(), seq.rend()));
    }
    printf("Answer to part 2 is %d\n", ans2);
    
    return 0;
}